File: src/main/java/leetcode/CombinationSum2.java
package leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Given a collection of candidate numbers (candidates) and a target number
 * (target), find all unique combinations in candidates where the candidate
 * numbers sum to target.
 *
 * Each number in candidates may only be used once in the combination.
 *
 * Note: The solution set must not contain duplicate combinations.
 */
final class CombinationSum2 {

    /**
     * Find all unique combinations in candidates where the candidate numbers
     * sum to target.
     *
     * The time complexity is O(2^N) where N is the length of the candidates,
     * because there are 2 choices at every iteration
     *
     * The space complexity is O(N) as there will be N recursive calls where N
     * is the length of candidates.
     *
     * @param candidates The input array
     * @param target     The target integer
     * @return The results
     */
    public List<List<Integer>> combinationSum2(final int[] candidates,
            final int target) {

        if (candidates.length == 0 || candidates.length > 101 || target < 1
                || target > 30) {
            return Collections.emptyList();
        }

        List<List<Integer>> result = new ArrayList<>();
        // sorting allows us to stop the DFS early
        Arrays.sort(candidates);
        backtrack(candidates, target, result, new ArrayList<>(), 0);
        return result;
    }

    /**
     * Helper method implementing the backtracking.
     *
     * @param candidates The input array
     * @param target     The target integer
     * @param result     The results array containing proven candidates
     * @param list       The proposed candidate
     * @param start      The starting index for iterating over the candidates
     */
    private void backtrack(final int[] candidates, final int target,
            final List<List<Integer>> result, final List<Integer> list,
            final int start) {

        if (target < 0) {
            // the candidate did not meet the conditions
            return;
        } else if (target == 0) {
            // the candidate did meet the conditions
            result.add(new ArrayList<>(list));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            // rely on sorting to avoid duplicates
            if (i == start || candidates[i] != candidates[i - 1]) {
                // simulate taking the number
                list.add(candidates[i]);
                // pass in i+1 as we can't add the current number again
                backtrack(candidates, target - candidates[i], result, list,
                        i + 1);
                // simulate not taking the number
                list.remove(list.size() - 1);
            }
        }
    }
}

File: python/src/combination_sum_2.py
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        """
        Given a collection of candidate numbers (candidates) and a target number (target), find all unique
        combinations in candidates where the candidate numbers sum to target.

        Each number in candidates may only be used once in the combination.

        Note: The solution set must not contain duplicate combinations.

        :param candidates: The list of candidates.
        :type candidates: List[int].
        :param target: The target number.
        :type target: Int.
        :return: The list of combinations that sum to the target.
        :rtype: List[List[int]].

        The time complexity is O(K * 2^N) where K is the average length of each solution and N is the length of the
        input list. Faster than 20.65% of solutions.

        The space complexity is O(N) for the recursion stack and to hold the list. This depends on the length of
        the longest solution which is equal to target/min of candidates. Less memory than 57.10% of solutions.
        """
        self.res = []
        self.candidates = candidates
        self.target = target

        # sort the array so that we can skip duplicates
        self.candidates.sort()
        self.backtrack(0, [], 0)

        return self.res

    def backtrack(self, pos, cur, total):
        """
        A helper method for backtracking.

        :param pos: Int.
        :type pos: The current position.
        :param cur: List.
        :type cur: The current sum.
        :param total: Int.
        :type total: The current total.
        :return: NoneType.
        :rtype: NoneType.
        """
        # base cases
        if total == self.target:
            self.res.append(cur.copy())
            return
        if total >= self.target:
            return

        prev_value = -1
        for i in range(pos, len(self.candidates)):
            # skip duplicates
            if self.candidates[i] == prev_value:
                continue
            cur.append(self.candidates[i])
            self.backtrack(i + 1, cur, total + self.candidates[i])
            cur.pop()
            prev_value = self.candidates[i]
